stlink: fix Darwin build

thefloweringash opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Make it build on Darwin.

The packaging assumes a static `.a` file for libusb, which nixpkgs does not provide. ~If we remove this special case for Darwin, it can successfully find libusb's `.dylib`.~

**Update**: make nixpkgs provide the static `.a` instead.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bjornfor commented:
Have you tried submitting the patch upstream?

Please explain stuff in the commit message, not only the PR message. (The latter does not become part of the git history.)

bjornfor commented:
Other than that (above), looks good.

thefloweringash committed:
libusb1: optionally enable static library

thefloweringash committed:
stlink: fix Darwin build

Upstream supports various Unixes, so optimistically set platforms to
`unix` even though we're only going to build it on Linux and Darwin.

thefloweringash commented:
I looked deeper at the build system in `stlink`, and it seems like the static library usage on Darwin is intentional and supports their binary releases. I think the easiest solution is to provide the static version of libusb (PR updated to do this). The nixpkgs documentation on `dontDisableStatic` doesn't give much of a guideline of when it's appropriate, but hopefully this is a valid use.
